sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUDF.scala

```diff
@@ -58,6 +58,8 @@ case class ScalaUDF(
 
   override lazy val deterministic: Boolean = udfDeterministic && children.forall(_.deterministic)
 
+  override lazy val foldable: Boolean = deterministic && children.forall(_.foldable)
```

Comment:
I think the problem we made ScalaUDF as deterministic by default so far, and we will have a lot of problems in existing workloads by executing the UDFs in driver side.

Reply:
If we ever made them non-determinstic by default, it would have been much easier to support foldability ..

Reply:
@HyukjinKwon 

> I think the problem we made ScalaUDF as deterministic by default so far, and we will have a lot of problems in existing workloads by executing the UDFs in driver side.

To me deterministic seems like the correct default, I would expect most udf to be deterministic. 

Seems like the concerns you raise are around the belief that people are using the existing API incorrectly. Do we have some more concrete evidence that this is the case? Because this seems like an argument that could be used against a lot of potential improvements and to me it does not seems like a rule we would like to follow in the long run.

My understanding it also that having a UDF with incorrect value for deterministic would already open you up to correctness issues due to other optimizer rules? So maybe that leads people to already specifying the correct nullability for their udfs.

Reply:
If user defines a UDF that requires something in executors, that'd be a problem now as the UDF runs in driver.

Reply:
Or, if the UDF runs something very heavy, it will stop the Catalyst optimizer.

Reply:
> Or, if the UDF runs something very heavy, it will stop the Catalyst optimizer.

To me this argument sounds a bit backwards. The heavier the UDF is the more gain it is from executing it once during planning instead of executing it multiple times during execution (worst case we are going to execute it once per row).


> If user defines a UDF that requires something in executors, that'd be a problem now as the UDF runs in driver.

Ok, this does sound like somthing that could cause breakage. Not sure how common that would be in practice b though. 


I changed the PR to make it opt in per UDF to make it foldable. I guess that should address most/all of the concerns you brought up with regards of it being a breaking change.

Reply:
@HyukjinKwon any chance we could modify this change to make it acceptable or should it just be abandoned?

I also notice that we will attempt to execute expressions and therefore UDFs as part of `EliminateOuterJoins`. There errors are ignored. We could extend the `ConstantFolding` rule to the same, ignore errors from expressions containing UDFs?